cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tether/TetherServerHandler.java

```diff
@@ -0,0 +1,254 @@
+      throw new NotImplementedException("Tethering is not enabled");
+    }
+  }
+
+  /**
+   * Sends control commands to the client.
+   */
+  @GET
+  @Path("/tethering/controlchannels/{peer}")
+  public void connectControlChannel(HttpRequest request, HttpResponder responder, @PathParam("peer") String peer)
+    throws IOException, NotImplementedException {
+    checkTetherServerEnabled();
+
+    store.updatePeer(peer, System.currentTimeMillis());
+    TetherStatus tetherStatus = store.getPeer(peer).getTetherStatus();
+    if (tetherStatus == TetherStatus.PENDING) {
+      responder.sendStatus(HttpResponseStatus.NOT_FOUND);
+      return;
+    } else if (tetherStatus == TetherStatus.REJECTED) {
+      responder.sendStatus(HttpResponseStatus.FORBIDDEN);
+      return;
+    }
+
+    List<TetherControlMessage> commands = new ArrayList<>();
+    MessageFetcher fetcher = messagingContext.getMessageFetcher();
+    TopicId topic = new TopicId(NamespaceId.SYSTEM.getNamespace(), TETHERING_TOPIC_PREFIX + peer);
+    String lastMessageId = null;
+    try (CloseableIterator<Message> iterator =
+           fetcher.fetch(topic.getNamespace(), topic.getTopic(), 1, lastMessageIds.get(topic.getTopic()))) {
+      while (iterator.hasNext()) {
+        Message message = iterator.next();
+        TetherControlMessage controlMessage = GSON.fromJson(message.getPayloadAsString(StandardCharsets.UTF_8),
+                                                            TetherControlMessage.class);
+        commands.add(controlMessage);
+        lastMessageId = message.getId();
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.warn("Received control connection from peer {} that's not tethered", peer);
+    }
+
+    if (lastMessageId != null) {
+      // Update the last message id for the topic if we read any messages
+      lastMessageIds.put(topic.getTopic(), lastMessageId);
+      String finalLastMessageId = lastMessageId;
+      TransactionRunners.run(transactionRunner, context -> {
+        AppMetadataStore.create(context).persistSubscriberState(topic.getTopic(), SUBSCRIBER, finalLastMessageId);
+      });
+    }
+
+    if (commands.isEmpty()) {
+      commands.add(new TetherControlMessage(TetherControlMessage.Type.KEEPALIVE, null));
+    }
+    Type type = new TypeToken<List<TetherControlMessage>>() { }.getType();
+    responder.sendJson(HttpResponseStatus.OK, GSON.toJson(commands, type));
+  }
+
+  /**
+   * Creates a tether with a client.
+   */
+  @POST
+  @Path("/tethering/connect")
+  public void createTether(FullHttpRequest request, HttpResponder responder)
+    throws NotImplementedException, IOException {
+    checkTetherServerEnabled();
+
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetherConnectionRequest tetherRequest = GSON.fromJson(content, TetherConnectionRequest.class);
+    TopicId topicId = new TopicId(NamespaceId.SYSTEM.getNamespace(), "tethering_" + tetherRequest.getPeer());
```

Comment:
nit: should use `TETHERING_TOPIC_PREFIX`

Reply:
fixed